### ISSUE TITLE
Improve AST

### ISSUE THREAD

**Issue by ethanuppal** [ethanuppal/pulsar#14]
- Provide a designated constructor for `Expr` and `Node`, hiding the token fields (and only exposing them through `TokenRegionProvider`).
- Add support for:
     - Type classes + struct types (dependent on #3)
     - Member access: properties and functions
     - Integration of type classes and builtin members on builtin types
- Clean up code (if possible, since parsers are guaranteed to be messy).

**Comment by ethanuppal** [comment]
Should `StmtType::make_unknown()` and related functions be moved to `StmtTypeCell` and appropriate paired types?

**Comment by ethanuppal** [comment]
The scope of this issue is too large if it subsumes #3, so I'm going to make that into a separate PR.
